### ISSUE TITLE
Jekyll incremental regenerations always renders .scss files, regardless of what file is modified?

### ISSUE THREAD

**Issue by bomberstudios** [jekyll/jekyll#4240]
I'm using Jekyll 3.0.1, and it seems to always be rendering .css files, regardless of what file I edit on the project.

I created a very simple demo project, with an `index.md` file with this code:

```markdown
---
---
# Hello, World
```

and a `style.scss` with this code:

```sass
---
---
$red: #f00;
div {
  color: $red;
}
```

Running `jekyll build --watch --verbose --incremental` I get this output:

```
  Logging at level: debug
Configuration file: none
            Source: /Users/ale/Documents/Code/jekyll-tests/scss-render-bug
       Destination: /Users/ale/Documents/Code/jekyll-tests/scss-render-bug/_site
 Incremental build: enabled
      Generating... 
       EntryFilter: excluded?(/index.md) ==> false
       EntryFilter: excluded?(/style.scss) ==> false
         Rendering: style.scss
  Pre-Render Hooks: style.scss
  Rendering Liquid: style.scss
  Rendering Markup: style.scss
 Post-Render Hooks: style.scss
                    done in 0.035 seconds.
           Watcher: Ignoring (?-mix:_site)
 Auto-regeneration: enabled for '/Users/ale/Documents/Code/jekyll-tests/scss-render-bug'
```

if I now edit `index.md`, I get this:

```
      Regenerating: 1 file(s) changed at 2015-12-09 08:55:58        EntryFilter: excluded?(/index.md) ==> false
       EntryFilter: excluded?(/style.scss) ==> false
         Rendering: index.md
  Pre-Render Hooks: index.md
  Rendering Liquid: index.md
  Rendering Markup: index.md
 Post-Render Hooks: index.md
         Rendering: style.scss
  Pre-Render Hooks: style.scss
  Rendering Liquid: style.scss
  Rendering Markup: style.scss
 Post-Render Hooks: style.scss
...done in 0.07034 seconds.
```

which is definitely *not* what I expected.

FTR, if I remove the frontmatter from the .scss file or rename it to `style.css` it works as expected (but then I can't use Sass syntax, which kinda defeats the purpose of the whole thing : )

Any ideas?

**Comment by envygeeks** [comment]
Logically, that is what's expected.  The dependency chain lines up IMO, you edit index.md, index.md uses layout.html, layout.html uses style.css, style.css is processed with layout.html with index.md.

**Comment by bomberstudios** [comment]
> you edit index.md, index.md uses layout.html, layout.html uses style.css

Check the example I provided. style.css is not used anywhere : )

**Comment by envygeeks** [comment]
> Check the example I provided. style.css is not used anywhere : )

What does it matter if I mispelled "style.scss" as "style.css" the comment does not change.

**Comment by envygeeks** [comment]
Unless you are implying there is no layout.html and layout.html does not have "style.scss" and if you are then I didn't grok that considering it didn't say "only the files without a layout" and it didn't say there is no layout anywhere because it doesn't imply any of that, sorry if I misinterpreted.

**Comment by bomberstudios** [comment]
Apologies accepted

**Comment by envygeeks** [comment]
I'm starting to get the feeling you are in a trolling mood so I'm going to ask directly instead of implying that you clarify your ticket and be very explicit about what your setup is so we can actually test it, if you cannot provide a repo please provide more detail as to the actual filesystem layout..

**Comment by bomberstudios** [comment]
If you could please read my original report, you'll see that everything is there, clearly explained. I will repeat that, using more words and pictures, just in case:

- I am using Jekyll 3.0.1. I installed it using `sudo gem install jekyll`, IIRC (I'm definitely **not** using Bundler in this demo)
- I made a simple demo project to replicate the issue, with only two files: `index.md` and `style.scss` (the contents of which are up there in the original report). There are no other files in the project. I'm attaching a screenshot of the folder, just in case it's not clear enough:

![screenshot 2015-12-09 12 19 57](https://cloud.githubusercontent.com/assets/3832/11684015/f74048cc-9e6e-11e5-9c27-acf9d866f4b7.png)

and this is my project in Terminal.app:

![screenshot 2015-12-09 12 22 20](https://cloud.githubusercontent.com/assets/3832/11684073/5c3bf640-9e6f-11e5-9098-d06c4bcc5e04.png)

If I run `jekyll build --watch --verbose --incremental` for the first time, (meaning, there's not a `_site` folder) I get this output:

```
  Logging at level: debug
Configuration file: none
            Source: /Users/ale/Documents/Code/jekyll-tests/scss-render-bug
       Destination: /Users/ale/Documents/Code/jekyll-tests/scss-render-bug/_site
 Incremental build: enabled
      Generating... 
       EntryFilter: excluded?(/index.md) ==> false
       EntryFilter: excluded?(/style.scss) ==> false
         Rendering: index.md
  Pre-Render Hooks: index.md
  Rendering Liquid: index.md
  Rendering Markup: index.md
 Post-Render Hooks: index.md
         Rendering: style.scss
  Pre-Render Hooks: style.scss
  Rendering Liquid: style.scss
  Rendering Markup: style.scss
 Post-Render Hooks: style.scss
                    done in 0.206 seconds.
           Watcher: Ignoring (?-mix:_site)
 Auto-regeneration: enabled for '/Users/ale/Documents/Code/jekyll-tests/scss-render-bug'
```

as expected. Now I'll edit **style.scss**. When I save it, I get this in Terminal.app:

```
      Regenerating: 1 file(s) changed at 2015-12-09 12:25:06        EntryFilter: excluded?(/index.md) ==> false
       EntryFilter: excluded?(/style.scss) ==> false
         Rendering: style.scss
  Pre-Render Hooks: style.scss
  Rendering Liquid: style.scss
  Rendering Markup: style.scss
 Post-Render Hooks: style.scss
...done in 0.007648 seconds.
```

Again, as expected, no surprises here.

If I now edit **index.md** (remember, this is just a Markdown file, with no layout, no linked CSS file, and absolutely no relationship *at all* with the other file in my project) and save it, I get this in Terminal.app:

```
      Regenerating: 1 file(s) changed at 2015-12-09 12:26:36        EntryFilter: excluded?(/index.md) ==> false
       EntryFilter: excluded?(/style.scss) ==> false
         Rendering: index.md
  Pre-Render Hooks: index.md
  Rendering Liquid: index.md
  Rendering Markup: index.md
 Post-Render Hooks: index.md
         Rendering: style.scss
  Pre-Render Hooks: style.scss
  Rendering Liquid: style.scss
  Rendering Markup: style.scss
 Post-Render Hooks: style.scss
...done in 0.009462 seconds.
```

Now **this** is unexpected. Jekyll itself is saying **1 file(s) changed**, yet it's rendering 2 files. This is the issue I'm reporting / asking for clarification.

Now, regarding your last comment: I believe the original report was clear enough, and I took the time to investigate the minimal set of conditions under which the issue occurs, hoping it would be appreciated and would help you pinpoint the issue / clarify what's going on. Instead, I'm being replied to in a snarky tone and being called a troll. That is *not cool*.

**Comment by envygeeks** [comment]
> If you could please read my original report, you'll see that everything is there, clearly explained.

No, it wasn't and if it was I wouldn't have made the assumptions I made and asked for more clarity later.
I am stepping out of this ticket and letting @jekyll/core take over, good luck.

**Comment by alfredxing** [comment]
I can repo this. It looks like `.jekyll-metadata` contains:

```json
{
  "/Users/alfred/Desktop/scss-render-bug/index.md": {
    "mtime": "2016-01-06 16:43:26 -0800",
    "deps": []
  }
}
```

Which indicates that the SCSS file is not being picked up by Regenerator at all...
I'll look into this further.

**Comment by bomberstudios** [comment]
Thanks a lot! : )

**Comment by alfredxing** [comment]
Okay, so after going through the history, it seems that the behaviour of having Sass/SCSS files always regenerated is forced, as discussed back in the original PR, starting from [this comment](https://github.com/jekyll/jekyll/pull/3116#issuecomment-65867834). See my and @parkr's follow-ups where we decide that asset files will always be regenerated.

**TLDR**: it's very hard or near impossible to keep track of dependencies within Sass, and since Sass generation is very quick, we decided to always regenerate Sass files.

**Comment by bomberstudios** [comment]
Thanks *so much* for the background info, much appreciated. Explanation makes sense, so I'm happy to close this : )

**Comment by parkr** [comment]
You're the best, @alfredxing! I totally forgot that.

**Comment by envygeeks** [comment]
Jumping back to that, why not add a tag that allows them to be partially tracked?

**Comment by envygeeks** [comment]
I ask because it's through the tag that we actually tie them into Jekyll's Regenerator so that it can inform us of an update to rebuild if we have a cache lying about.

**Comment by alfredxing** [comment]
:+1:

**Comment by alfredxing** [comment]
@envygeeks Here's my thinking/logic on this, let me know if it doesn't make sense:
- The dependency between Sass files and documents that include them isn't quite the kind that Regenerator is tracking, since a change in the Sass file does not directly affect the content of the referencing document as generated by Jekyll
- Another issue here (raised in the comment on the original PR I linked above) is that even if we track the "main" Sass file, changes in Sass includes (which are a part of the Sass language themselves and not something we control) will not cause the main Sass file to be regenerated, since the main file and _its_ dependencies aren't tracked.

**Comment by envygeeks** [comment]
* Point 1: It depends on how you work with SASS and assets in general.  With Jekyll's simplistic approach to SASS, you are right, it doesn't matter, but when more complex scenarios happen (for example when somebody generates a random bust) it becomes beneficial.  The tracking here is not for your own sake, it's for the users corner

* Point 2: It's easy to track SASS dependencies because it's Ruby, and it must have it's own middleware to deal with that (since itself has caching) so you can easily tie into that the way that Sprockets and Rails do, and dozens of other SASS products.

**Comment by envygeeks** [comment]
The lot of this though, is not relevant to Jekyll itself though, it belongs on jekyll-sass.
